src/ioctls/common/kvm_vec.rs

```diff
@@ -0,0 +1,549 @@
+        let desired_capacity = self.len + additional;
+        if desired_capacity <= self.capacity {
+            return;
+        }
+
+        let current_mem_allocator_len = self.mem_allocator.len();
+        let required_mem_allocator_len =
+            KvmVec::<T>::kvm_vec_len_to_mem_allocator_len(desired_capacity);
+        let additional_mem_allocator_len = required_mem_allocator_len - current_mem_allocator_len;
+
+        self.mem_allocator.reserve(additional_mem_allocator_len);
+        self.capacity =
+            KvmVec::<T>::mem_allocator_len_to_kvm_vec_len(self.mem_allocator.capacity());
+    }
+
+    /// Updates the length of `self` to the specified value.
+    /// Also updates the length of the `T::Entry` structure and of `self.mem_allocator` accordingly.
+    ///
+    fn update_len(&mut self, len: usize) {
+        self.len = len;
+        self.as_mut_kvm_struct().set_len(len);
+
+        /// We need to set the len of the mem_allocator to be the number of T elements needed
+        /// to fit an array of `len` elements of type `T::Entry`. This way, when we call
+        /// `self.mem_allocator.shrink_to_fit()` only the unnecessary memory will be released.
+        let required_mem_allocator_len = KvmVec::<T>::kvm_vec_len_to_mem_allocator_len(len);
+        unsafe {
+            self.mem_allocator.set_len(required_mem_allocator_len);
+        }
+    }
+
+    /// Appends an element to the end of the collection and updates `len`.
+    ///
+    /// # Arguments
+    ///
+    /// * `entry` - The element that will be appended to the end of the collection.
+    ///
+    /// # Error: When len is already equal to max possible len it returns Error::SizeLimitExceeded
+    ///
+    pub fn push(&mut self, entry: T::Entry) -> Result<(), Error> {
+        let desired_len = self.len + 1;
+        if desired_len > T::max_len() {
+            return Err(Error::SizeLimitExceeded);
+        }
+
+        self.reserve(1);
+
+        let mut entries = self.as_vec();
+        entries.push(entry);
+        self.update_len(desired_len);
+
+        mem::forget(entries);
+
+        Ok(())
+    }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// # Arguments
+    ///
+    /// * `f` - The function used to evaluate whether an entry will be kept or not.
+    ///         When `f` returns `true` the entry is kept.
+    ///
+    pub fn retain<P>(&mut self, f: P)
+    where
+        P: FnMut(&T::Entry) -> bool,
+    {
+        let mut entries = self.as_vec();
+        entries.retain(f);
+
+        self.update_len(entries.len());
+        self.mem_allocator.shrink_to_fit();
+        self.capacity =
+            KvmVec::<T>::mem_allocator_len_to_kvm_vec_len(self.mem_allocator.capacity());
+
+        mem::forget(entries);
+    }
+}
+
+impl<T: Default + KvmArray> PartialEq for KvmVec<T> {
+    fn eq(&self, other: &KvmVec<T>) -> bool {
+        self.len == other.len && self.as_entries_slice() == other.as_entries_slice()
```

Comment:
Is `self.len == other.len` implied by `self.as_entries_slice() == other.as_entries_slice()`?